File: HomeWork05/legendre_stability.py
from decimal import Decimal, getcontext
from pathlib import Path
import csv
import math

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt


LMAX = 50
L_BACK = 80
X_VALUES = [0.1, 0.5, 0.9, 0.99]
PRECISION_DIGITS = 110


def legendre_forward_double(x, lmax):
    values = [0.0] * (lmax + 1)
    values[0] = 1.0

    if lmax >= 1:
        values[1] = x

    for ell in range(1, lmax):
        values[ell + 1] = (
            ((2.0 * ell + 1.0) / (ell + 1.0)) * x * values[ell]
            - (ell / (ell + 1.0)) * values[ell - 1]
        )

    return values


def legendre_reference_decimal(x_string, lmax):
    x = Decimal(x_string)
    values = [Decimal(0)] * (lmax + 1)
    values[0] = Decimal(1)

    if lmax >= 1:
        values[1] = x

    for ell in range(1, lmax):
        ell_d = Decimal(ell)
        values[ell + 1] = (
            (Decimal(2 * ell + 1) / Decimal(ell + 1)) * x * values[ell]
            - (ell_d / Decimal(ell + 1)) * values[ell - 1]
        )

    return values


def legendre_backward_double(x, lmax, start_l):
    if start_l <= lmax:
        raise ValueError("start_l must be larger than lmax")

    q = [0.0] * (start_l + 2)
    q[start_l + 1] = 0.0
    q[start_l] = 1.0

    for ell in range(start_l, 0, -1):
        q[ell - 1] = (
            ((2.0 * ell + 1.0) / ell) * x * q[ell]
            - ((ell + 1.0) / ell) * q[ell + 1]
        )

    scale = 1.0 / q[0]
    return [scale * q[ell] for ell in range(lmax + 1)]


def decimal_abs(value):
    return value.copy_abs()


def errors(numerical, reference):
    value = Decimal.from_float(numerical)
    abs_err = decimal_abs(value - reference)
    if decimal_abs(reference) > Decimal("1e-90"):
        rel_err = abs_err / decimal_abs(reference)
    else:
        rel_err = abs_err
    return abs_err, rel_err


def as_float_for_plot(value):
    numeric = float(value)
    if numeric == 0.0:
        return 1e-35
    return numeric


def compute_rows():
    getcontext().prec = PRECISION_DIGITS
    rows = []

    for x in X_VALUES:
        x_string = format(x, ".17g")
        reference = legendre_reference_decimal(x_string, LMAX)
        forward = legendre_forward_double(x, LMAX)
        backward = legendre_backward_double(x, LMAX, L_BACK)

        for ell in range(LMAX + 1):
            abs_f, rel_f = errors(forward[ell], reference[ell])
            abs_b, rel_b = errors(backward[ell], reference[ell])

            rows.append(
                {
                    "x": x,
                    "ell": ell,
                    "P_reference": str(reference[ell]),
                    "P_forward": format(forward[ell], ".17g"),
                    "P_backward": format(backward[ell], ".17g"),
                    "abs_err_forward": str(abs_f),
                    "rel_err_forward": str(rel_f),
                    "abs_err_backward": str(abs_b),
                    "rel_err_backward": str(rel_b),
                }
            )

    return rows


def write_csv(rows, path):
    fieldnames = [
        "x",
        "ell",
        "P_reference",
        "P_forward",
        "P_backward",
        "abs_err_forward",
        "rel_err_forward",
        "abs_err_backward",
        "rel_err_backward",
    ]

    with path.open("w", newline="") as file:
        writer = csv.DictWriter(file, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(rows)


def plot_error(rows, column, ylabel, title, path):
    plt.figure(figsize=(7, 4.5))

    for x in X_VALUES:
        selected = [row for row in rows if math.isclose(row["x"], x)]
        ell = [row["ell"] for row in selected]
        err = [as_float_for_plot(Decimal(row[column])) for row in selected]
        plt.plot(ell, err, marker="o", markersize=3, linewidth=1, label=f"x = {x:g}")

    plt.yscale("log")
    plt.xlabel("degree ell")
    plt.ylabel(ylabel)
    plt.title(title)
    plt.grid(True, which="both", alpha=0.3)
    plt.legend()
    plt.tight_layout()
    plt.savefig(path, dpi=180)
    plt.close()


def print_summary(rows):
    print("HomeWork05 - Stability of Legendre Polynomials")
    print(f"lmax = {LMAX}, backward start L = {L_BACK}")
    print()

    for x in X_VALUES:
        selected = [row for row in rows if math.isclose(row["x"], x)]
        max_forward = max(Decimal(row["rel_err_forward"]) for row in selected)
        max_backward = max(Decimal(row["rel_err_backward"]) for row in selected)
        print(
            f"x = {x:g}: max rel error forward = {max_forward:.3E}, "
            f"backward = {max_backward:.3E}"
        )


def main():
    base = Path(__file__).resolve().parent
    plots_dir = base / "plots"
    plots_dir.mkdir(exist_ok=True)

    rows = compute_rows()
    write_csv(rows, base / "legendre_errors.csv")

    plot_error(
        rows,
        "rel_err_forward",
        "relative error",
        "Forward recurrence: relative error",
        plots_dir / "forward_relative_error.png",
    )
    plot_error(
        rows,
        "rel_err_backward",
        "relative error",
        "Backward recurrence experiment: relative error",
        plots_dir / "backward_relative_error.png",
    )
    plot_error(
        rows,
        "abs_err_forward",
        "absolute error",
        "Forward recurrence: absolute error",
        plots_dir / "forward_absolute_error.png",
    )
    plot_error(
        rows,
        "abs_err_backward",
        "absolute error",
        "Backward recurrence experiment: absolute error",
        plots_dir / "backward_absolute_error.png",
    )

    print_summary(rows)
    print()
    print("Wrote legendre_errors.csv and plots/*.png")


if __name__ == "__main__":
    main()

File: HomeWork05/legendre_stability.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#define LMAX 50
#define L_BACKWARD 80
#define N_X 4

static const double x_values[N_X] = {0.1, 0.5, 0.9, 0.99};

typedef struct {
    double hi;
    double lo;
} dd_t;

static FILE *open_file(const char *filename)
{
    FILE *file = fopen(filename, "w");

    if (file == NULL) {
        perror(filename);
        exit(EXIT_FAILURE);
    }

    return file;
}

static dd_t dd_normalize(double hi, double lo)
{
    dd_t result;
    double sum = hi + lo;
    double err = lo - (sum - hi);

    result.hi = sum;
    result.lo = err;
    return result;
}

static dd_t dd_from_double(double value)
{
    dd_t result = {value, 0.0};

    return result;
}

static dd_t dd_add(dd_t a, dd_t b)
{
    double sum = a.hi + b.hi;
    double v = sum - a.hi;
    double err = (a.hi - (sum - v)) + (b.hi - v) + a.lo + b.lo;

    return dd_normalize(sum, err);
}

static dd_t dd_sub(dd_t a, dd_t b)
{
    dd_t minus_b = {-b.hi, -b.lo};

    return dd_add(a, minus_b);
}

static dd_t dd_mul_double(dd_t a, double b)
{
    double product = a.hi * b;
    double err = fma(a.hi, b, -product) + a.lo * b;

    return dd_normalize(product, err);
}

static double dd_to_double(dd_t a)
{
    return a.hi + a.lo;
}

static double dd_abs(dd_t a)
{
    return fabs(dd_to_double(a));
}

/*
 * Forward Bonnet recurrence in double precision:
 *
 *     P_0(x) = 1
 *     P_1(x) = x
 *     P_{l+1}(x) = ((2l+1)/(l+1)) x P_l(x) - (l/(l+1)) P_{l-1}(x)
 */
static void legendre_forward_double(double x, double p[LMAX + 1])
{
    p[0] = 1.0;
    p[1] = x;

    for (int l = 1; l < LMAX; ++l) {
        p[l + 1] = ((2.0 * l + 1.0) / (l + 1.0)) * x * p[l]
                 - (l / (l + 1.0)) * p[l - 1];
    }
}

/*
 * Forward recurrence in double-double precision.
 * This provides a C reference more accurate than ordinary double arithmetic.
 */
static void legendre_reference_dd(double x, dd_t p[LMAX + 1])
{
    p[0] = dd_from_double(1.0);
    p[1] = dd_from_double(x);

    for (int l = 1; l < LMAX; ++l) {
        double coeff1 = ((2.0 * l + 1.0) / (l + 1.0)) * x;
        double coeff2 = l / (l + 1.0);
        dd_t term1 = dd_mul_double(p[l], coeff1);
        dd_t term2 = dd_mul_double(p[l - 1], coeff2);

        p[l + 1] = dd_sub(term1, term2);
    }
}

/*
 * Backward recurrence experiment.
 *
 * We start from arbitrary values at high degree:
 *
 *     Q_{L+1} = 0
 *     Q_L     = 1
 *
 * Then we recurse downward and rescale so that Q_0 = 1.
 */
static void legendre_backward_double(double x, double p[LMAX + 1])
{
    double q[L_BACKWARD + 2];

    for (int l = 0; l <= L_BACKWARD + 1; ++l) {
        q[l] = 0.0;
    }

    q[L_BACKWARD + 1] = 0.0;
    q[L_BACKWARD] = 1.0;

    for (int l = L_BACKWARD; l > 0; --l) {
        q[l - 1] = ((2.0 * l + 1.0) / l) * x * q[l]
                 - ((l + 1.0) / l) * q[l + 1];
    }

    for (int l = 0; l <= LMAX; ++l) {
        p[l] = q[l] / q[0];
    }
}

static double relative_error(dd_t value, dd_t reference)
{
    double abs_error = dd_abs(dd_sub(value, reference));
    double ref_abs = dd_abs(reference);

    if (ref_abs > 1.0e-90) {
        return abs_error / ref_abs;
    }

    return abs_error;
}

int main(void)
{
    FILE *file = open_file("legendre_errors.csv");

    fprintf(file,
            "x,ell,P_reference,P_forward,P_backward,"
            "abs_err_forward,rel_err_forward,"
            "abs_err_backward,rel_err_backward\n");

    for (int ix = 0; ix < N_X; ++ix) {
        double x = x_values[ix];
        double p_forward[LMAX + 1];
        double p_backward[LMAX + 1];
        dd_t p_reference[LMAX + 1];
        double max_rel_forward = 0.0;
        double max_rel_backward = 0.0;

        legendre_forward_double(x, p_forward);
        legendre_backward_double(x, p_backward);
        legendre_reference_dd(x, p_reference);

        for (int l = 0; l <= LMAX; ++l) {
            dd_t ref = p_reference[l];
            dd_t fwd = dd_from_double(p_forward[l]);
            dd_t bwd = dd_from_double(p_backward[l]);
            double abs_err_forward = dd_abs(dd_sub(fwd, ref));
            double abs_err_backward = dd_abs(dd_sub(bwd, ref));
            double rel_err_forward = relative_error(fwd, ref);
            double rel_err_backward = relative_error(bwd, ref);

            if (rel_err_forward > max_rel_forward) {
                max_rel_forward = rel_err_forward;
            }
            if (rel_err_backward > max_rel_backward) {
                max_rel_backward = rel_err_backward;
            }

            fprintf(file,
                    "%.17g,%d,%.17g,%.17g,%.17g,"
                    "%.17e,%.17e,%.17e,%.17e\n",
                    x, l, dd_to_double(ref), p_forward[l], p_backward[l],
                    abs_err_forward, rel_err_forward,
                    abs_err_backward, rel_err_backward);
        }

        printf("x = %.17g: max relative error forward = %.3e, "
               "backward = %.3e\n",
               x, max_rel_forward, max_rel_backward);
    }

    fclose(file);
    printf("Wrote legendre_errors.csv\n");

    return 0;
}
